File: quicksort/quick_sort.py
def median(a, i, j, k):
    """
    Returns Median of 3 integers from array a.
    :param a: Iterable of elements
    :param i: index of element of set, whose meadian is to be returned
    :param j: index of element of set, whose meadian is to be returned
    :param k: index element of set, whose meadian is to be returned
    :return: returns the value of index median of a[i], a[j], a[k].
    """
    if a[i] > a[j]:
        return j if a[j] < a[k] else k
    else:
        return i if a[i] < a[k] else k


def partition(array, l, r):
    """
    Perform Partition Operation on array.
    Time Complexity: Theta(nLogn)
    Auxiliary Space: O(n)
    :param array: Iterable of elements
    :param l: left limit of array
    :param r: right limit of array
    :return: return q (partition index) value for function, used in partitioning of array.
    """
    i = l - 1
    pivot_index = median(array, l, r, (l + r) // 2)
    pivot = array[pivot_index]
    for j in range(l, r+1):
        if array[j] <= pivot:
            i += 1
            array[i], array[j] = array[j], array[i]
    i += 1
    array[i], array[r] = array[r], array[i]
    return i


def quick_sort(array, pivot, right):
    """
    Perform sort using partition function.
    Time Complexity : O(nlog(n)).
    Space Complexity : O(n).
    :param array: Iterable of elements
    :param pivot: used as left limit of quick sort
    :param right: right limit for quick sort
    :return: no returns, sorts array
    """
    if pivot < right:
        q = partition(array, pivot, right)
        quick_sort(array, pivot, q - 1)
        quick_sort(array, q, right)


def main():
    a = [15, 19, 18, 26, 456, 87, 45, 480, 5]
    quick_sort(a, 0, len(a) - 1)
    print(a)

if __name__ == '__main__':
    main()

File: quicksort/quick_sort.go
package main

import "fmt"

// partitionArray finds the pivot index
func partitionArray(data []int, left int, right int) int {
	pivotIndex := left
	for true {
		for data[pivotIndex] <= data[right] && pivotIndex != right {
			right--
		}
		if pivotIndex == right {
			break
		} else if data[pivotIndex] > data[right] {
			data[right], data[pivotIndex] = data[pivotIndex], data[right]
			pivotIndex = right
		}
		for data[pivotIndex] >= data[left] && pivotIndex != left {
			left++
		}
		if pivotIndex == left {
			break
		} else if data[pivotIndex] < data[left] {
			data[left], data[pivotIndex] = data[pivotIndex], data[left]
			pivotIndex = left
		}
	}
	return pivotIndex
}

func quickSort(data []int, begin int, end int) {
	if begin < end {
		pivotIndex := partitionArray(data, begin, end)
		quickSort(data, begin, pivotIndex-1)
		quickSort(data, pivotIndex+1, end)
	}
}

// QuickSort sorts data by quicksort algorithm
// Time complexity : O(n log n)
// Space Complexity : O(n)
func QuickSort(data []int) {
	quickSort(data, 0, len(data)-1)
}

func main() {
	data := []int{1, 1122002, 2, 88171, 6754, 79901, 119856, -312, 1, -2}
	QuickSort(data)
	fmt.Println(data)
}
